### PR TITLE
Add support for total patterns in lambdas

### LINKED ISSUE #188: support patterns in lambdas
We should support irrefutable patterns in lambdas:
```
f = \(x: Int) -> x.add(1)

f = \(x, y) -> x.add(y)

const = \_ -> 42

struct Order(fn: a -> a -> Comparison)
f = \Order(fn), lst -> lst.sort(fn)
```

This would have made the curry examples slightly cleaner: #185

### PR DESCRIPTION
**Pull request by johnynek**
close #188 

This will be pretty nice in a lot of cases, is easy to support.

What do you think @snoble ?

### CODE REVIEW

File: core/src/test/scala/org/bykn/bosatsu/EvaluationTest.scala [hunk 1]
```diff
@@ -1086,4 +1086,33 @@ main = Foo(1, `package`, 3, 4)
...
+inc = \Foo(x) -> x.add(1)
+
+test0 = Assertion(inc(Foo(1)).eq_Int(2), "inc(Foo(1)) == 2")
+
+enum FooBar: F(x), B(x)
+
+inc2 = \F(x) | B(x), Foo(y) -> x.add(y)
+test1 = Assertion(inc2(F(1), Foo(1)).eq_Int(2), "inc2(F(1), Foo(1)) == 2")
+test2 = Assertion(inc2(B(1), Foo(1)).eq_Int(2), "inc2(B(1), Foo(1)) == 2")
+
```
Comment (snoble) [inline core/src/test/scala/org/bykn/bosatsu/EvaluationTest.scala:1113]: This is really cool!

Maybe add a couple tests for a multi-parameter struct and a nested pattern? Like `inc3 = \(F(x) | B(x), Foo(y)) -> x.add(y)` for example

File: core/src/main/scala/org/bykn/bosatsu/Expr.scala [hunk 2]
```diff
@@ -177,5 +177,27 @@ object Expr {
         buildLambda(NonEmptyList.of(arg), body1, outer)
     }
 
+  def buildPatternLambda[A](
```
Comment (snoble) [inline core/src/main/scala/org/bykn/bosatsu/Expr.scala:180]: Looks like `buildLambda` is only used by `DefStatement` now. So not quite ready to delete it... but soon maybe?

Comment (johnynek) [reply core/src/main/scala/org/bykn/bosatsu/Expr.scala:180]: this is interesting. Should we allow patterns in defs?
```
def getBar(Bar(x)): x
```
doesn't look much like python, but if you want to go nuts, Haskell has something like this that allows multiple branches:
```
enum Either: Left(a), Right(b)

def map(Right(a), fn): Right(fn(a))
def map(left @ Left(_), _): left
```
I'm not crazy about that syntax. I think a match would be better.

Comment (snoble) [reply core/src/main/scala/org/bykn/bosatsu/Expr.scala:180]: My first thought when I see this is someone could get the impression that bosatsu has function overloading. And I think a match is always more clear

But for the non enum case this seems really natural and would be confusing if it were absent I think

Comment (johnynek) [reply core/src/main/scala/org/bykn/bosatsu/Expr.scala:180]: Consistency is a powerful argument. Okay. I think I agree with your take: yes to patterns in defs, no to multi defs for enums. So harmony with lambdas.

**Pull request merged**